contracts/src/VouchMe.sol

```diff
@@ -18,68 +17,112 @@ contract VouchMe is ERC721URIStorage {
     // Maps token ID to testimonial details
     mapping(uint256 => Testimonial) private _testimonials;
     
-    struct Testimonial {
+    // Maps sender to receiver to their active testimonial (for one-per-pair rule)
+    mapping(address => mapping(address => uint256)) private _activeTestimonial;
+    
+    // Maps user address to their profile data
+    mapping(address => Profile) public userProfiles;    
+    struct Profile {
+        string name;
+        string contact;
+        string bio;
+    }
+      struct Testimonial {
         address sender;
         address receiver;
         string content;
+        string giverName;   
+        string profileUrl;    
         uint256 timestamp;
         bool verified;
+        bool isDeleted;
     }
     
     event TestimonialCreated(uint256 tokenId, address sender, address receiver);
     event TestimonialVerified(uint256 tokenId, address receiver);
+    event TestimonialRemoved(uint256 tokenId, address receiver);
+    event TestimonialUpdated(address sender, address receiver, uint256 newTokenId);
+    event ProfileUpdated(address user);
     
-    constructor() ERC721("VouchMe Testimonial", "VOUCH") {}
-
-    /**
+    constructor() ERC721("VouchMe Testimonial", "VOUCH") {}    /**
      * @dev Creates a testimonial NFT based on a signed message
      * @param senderAddress Address of the sender who created the testimonial
      * @param content The testimonial content
+     * @param giverName Full name of the person giving the testimonial
+     * @param profileUrl Optional LinkedIn or GitHub profile URL (can be empty)
      * @param signature Signature of the testimonial data
      * @return tokenId The ID of the newly created testimonial NFT
      */
     function createTestimonial(
         address senderAddress, 
-        string calldata content, 
+        string calldata content,
+        string calldata giverName,
+        string calldata profileUrl,
         bytes calldata signature
-    ) external returns (uint256) {
-        // Hash the message that was signed
+    ) external returns (uint256) {        // Hash the message that was signed
         bytes32 messageHash = keccak256(
             abi.encodePacked(
                 senderAddress,
                 msg.sender, // receiver
-                content
+                content,
+                giverName,
+                profileUrl
             )
         );
         
         bytes32 ethSignedMessageHash = MessageHashUtils.toEthSignedMessageHash(messageHash);
-        
-        // Verify the signature matches the sender
+          // Verify the signature matches the sender
         address recoveredSigner = ethSignedMessageHash.recover(signature);
         require(recoveredSigner == senderAddress, "Invalid signature");
 
+        // Check if there's an existing active testimonial from this sender to this receiver
+        uint256 existingTokenId = _activeTestimonial[senderAddress][msg.sender];
+        if (existingTokenId != 0) {
```

Comment:
This if block appears to duplicate code from `removeTestimonial`. Couldn't we refactor the code to eliminate the duplication?